lib/include/jxl/decode.h

```diff
@@ -913,6 +953,144 @@ JXL_EXPORT JxlDecoderStatus JxlDecoderSetJPEGBuffer(JxlDecoder* dec,
  */
 JXL_EXPORT size_t JxlDecoderReleaseJPEGBuffer(JxlDecoder* dec);
 
+/**
+ * Sets output buffer for box output codestream.
+ *
+ * The data is owned by the caller and may be used by the decoder until
+ * JxlDecoderReleaseBoxBuffer is called or the decoder is destroyed or reset so
+ * must be kept alive until then.
+ *
+ * @param dec decoder object
+ * @param data pointer to next bytes to write to
+ * @param size amount of bytes available starting from data
+ * @return JXL_DEC_ERROR if output buffer was already set and
+ * JxlDecoderReleaseBoxBuffer was not called on it, JXL_DEC_SUCCESS otherwise
+ */
+JXL_EXPORT JxlDecoderStatus JxlDecoderSetBoxBuffer(JxlDecoder* dec,
+                                                   uint8_t* data, size_t size);
+
+/**
+ * Releases buffer which was provided with JxlDecoderSetBoxBuffer.
+ *
+ * Calling JxlDecoderReleaseBoxBuffer is required whenever
+ * a buffer is already set and a new buffer needs to be added with
+ * JxlDecoderSetBoxBuffer, but is not required before JxlDecoderDestroy or
+ * JxlDecoderReset.
+ *
+ * Calling JxlDecoderReleaseBoxBuffer when no buffer is set is
+ * not an error and returns 0.
+ *
+ * @param dec decoder object
+ * @return the amount of bytes the decoder has not yet written to of the data
+ * set by JxlDecoderSetBoxBuffer, or 0 if no buffer is set or
+ * JxlDecoderReleaseBoxBuffer was already called.
+ */
+JXL_EXPORT size_t JxlDecoderReleaseBoxBuffer(JxlDecoder* dec);
+
+/**
+ * Configures whether to get boxes in raw mode or in uncompressed
+ * mode.
+ *
+ * The default mode is non-raw (JXL_FALSE), where "brob" boxes are transparently
+ * decompressed. This setting can only be changed
+ * before decoding, or directly after a JXL_DEC_BOX event, and is preserved
+ * otherwise.
+ *
+ * @param dec decoder object
+ * @param enable_raw whether to enable or disable raw mode
+ * @return JXL_DEC_ERROR if input was already set without releasing,
+ * JXL_DEC_SUCCESS otherwise
+ */
+JXL_EXPORT void JxlDecoderSetBoxModeRaw(JxlDecoder* dec, JXL_BOOL enable_raw);
+
+/**
+ * Outputs the type of the current box, after a JXL_DEC_BOX event occured, as 4
+ * characters plus a null termination character. In case of a compressed box,
+ * this returns the type of the underlying uncompressed box, not "brob". Use
+ * @see JxlDecoderGetBoxTypeRaw to get the raw type.
+ *
+ * @param dec decoder object
+ * @param type buffer to copy the type into
+ * @return JXL_DEC_SUCCESS if the value is available, JXL_DEC_ERROR if not, for
+ *    example the JXL file does not use the container format.
+ */
+JXL_EXPORT JxlDecoderStatus JxlDecoderGetBoxType(JxlDecoder* dec, char type[5]);
```

Comment:
I would propose to use a `typedef struct { char box_type[4]; } JxlBoxType` instead of the raw char type[] that's sometimes 5 bytes and sometimes 4 bytes.

Reply:
I've done this but with `typedef char JxlBoxType[4]`

This solves the ambiguity of 4 vs 5. But without struct, it also prevents double naming like "box_type.box_type"

Reply:
> It might be useful to add a (currently hypothetical) example application that uses this new API. For example an application that takes a jxl and dumps the XMP metadata to a file. Or an application that takes a jxl and produces a new jxl where all the compressed boxes are uncompressed (say so they can be processed by 'dumb' metadata tools).
> 
> I think the API as it is described here is OK, but it would be instructive to have an example application to see if we're not missing something.
> 
> Also (not for this pull request, obviously), I think (as we have discussed before) we should add a simpler high-level API that doesn't do things in the streaming+events way but that you just give a filename or memory pointer and gives you an image buffer + uncompressed metadata blobs without you having to worry about memory allocation and all that. For simple applications, it would be good to have a convenient high-level one-shot API that you can use very easily, something like
> 
> ```
> char *filename;
> uint8_t *display_profile; // points to an ICC profile describing the display space, or nullptr for sRGB
> uint8_t *rgba; // will contain 8-bit interleaved RGBA data, possibly multiple frames 
> JxlBlobs *blobs;  // struct that contains a list of tuples of (blobtype, data, data_size)
> size_t w, h;
> JxlAnimInfo *animinfo; // struct that contains nb of frames and frame duration
> 
> JxlRetCode = JxlSimpleDecodeRGBA8(filename, display_profile, rgba, blobs, &w, &h, animinfo);
> if (JxlRetCode != JxlOK) { ... } // oops, something went wrong
> if (animinfo->nb_frames > 1) { ... } // render as an animation
> // yay we have an image we can put on the screen
> JxlBlob exif = JxlGetBlob(blobs, "Exif");
> if (exif.size > 0) {
>   MyViewerAppShowExifMetadata(exif.data, exif.size);
> }
> ```

Adding an example is a good idea, done!

I have not yet integrated the example in the build script because it of course can't yet link